c/compare.py

```diff
@@ -0,0 +1,174 @@
+    print("Skipping category", setf, "(only custom includes, no system headers, complicated build process)")
+    continue
+  elif setf == "Systems_SQLite_MemSafety":
+    print("Skipping category", setf, "(complicated build process, requires patched version of cilly)")
+    continue
+
+  if not os.path.exists(setf + ".cfg"):
+    print("Skipping category", setf, "(no .cfg file present)")
+    continue
+
+  print("Processing category", setf)
+  bits=getArchitecture(setf + ".cfg")
+  if bits not in ["32", "64"]:
+    print("Invalid bit width in file", setf + ".cfg")
+    exit(1)
+
+  # iterate over all files in the set
+  i=0
+  for ff in getTasksFromSet(f):
+    orig=ff
+
+    if ff.endswith(".yml"):
+      with open(ff, 'r') as yamlfile:
+        yml = yaml.safe_load(yamlfile)
+
+      # check whether the input_basename exists (nobody should ever use "null" as a filename!)
+      inputFiles = yml['input_files']
+      if not inputFiles:
+        print("No input files defined in", ff)
+        exit(1)
+
+      # check whether there is exactly one input file, either directly or nested in a list
+      if isinstance(inputFiles, list):
+        print("ignoring task consisting of multiple sourcefiles")
+        continue
+      ff = os.path.join(os.path.dirname(ff), inputFiles)
+
+    if not os.path.exists(ff):
+      print("No task file", ff, "found")
+      exit(1)
+
+    # no original source available
+    if ff.startswith(('ddv-machzwd/', 'aws-c-common/', 'ldv-linux-3.0/', 'ldv-regression/')):
+      # for LDV: there is a related .cil.c file, but it doesn't necessarily match at all
+      continue
+    if ff == "loops/s3.i":
+      continue
+
+    # try to find a matching source file for a preprocessed task
+    if ff.endswith('.c'):
+      continue
+    elif ff.endswith('.c.i'):
+      orig = ff[:-2] # remove ending ".i"
+    elif ff.startswith('ldv-memsafety/memleaks'):
+      orig = ff.replace('memleaks', 'memleaks-notpreprocessed/memleaks')
+      orig = orig[:-2] + ".c" # replace .i with .c
+    else:
+      orig = ff[:-2] + ".c" # replace .i with .c
+
+    if not os.path.exists(orig):
+      print("No original source of", ff, "found")
+      exit(1)
+
+    i += 1
+    if i % 10 == 0:
+      print("Processing file", i, "of category", setf)
+
+    # now we have found all required files and start with the actual check:
+    # convert both files into goto-cc intermediate language and compare them.
+    subprocess.call(["goto-cc", "-m" + bits, orig])
+    subprocess.call(["goto-cc", "-m" + bits, ff, "-o", "b.out"])
+    p = subprocess.Popen(["goto-diff", "--verbosity", "2", "-u", "a.out", "b.out"], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    stdout, stderr = p.communicate()
+    if len(stderr) > 0:
+        print(stderr.decode('utf-8').strip())
+    if len(stdout) > 0:
+      if options.SHOW_DIFF:
+        subprocess.call(["goto-diff", "-u", "a.out", "b.out"])
+      shutil.rmtree("a.out", ignore_errors=True)
```

Comment:
I would use `shutil.rmtree` only if we actually need to delete a tree. It is just too easy to make a mistake and accidentally delete lots of user data (for example, if the script is called while `a.out` is a directory). `os.remove` just deletes a single file.